lib/blemulator/internal/simulation_manager.dart

```diff
@@ -24,17 +24,92 @@ class SimulationManager {
 
   Future<void> _startDeviceScan() async {
     _peripherals.forEach((peripheral) {
-      _scanSubscriptions.add(peripheral
-          .onScan(allowDuplicates: true)
-          .listen((scanResult) {
-            print(scanResult);
-            return _bridge.addScanResult(scanResult);
-          }));
+      _scanSubscriptions
+          .add(peripheral.onScan(allowDuplicates: true).listen((scanResult) {
+        print(scanResult);
+        return _bridge.publishScanResult(scanResult);
+      }));
     });
   }
 
   Future<void> _stopDeviceScan() async {
     _scanSubscriptions.forEach((subscription) => subscription.cancel());
     _scanSubscriptions.clear();
   }
+
+  Future<SimulatedCharacteristic> _readCharacteristicForIdentifier(
+    int characteristicIdentifier,
+  ) {
+    SimulatedCharacteristic targetCharacteristic;
+    peripheralsLoop:
+    for (SimulatedPeripheral peripheral in _peripherals) {
```

Comment:
1) I propose `SimulationManager.getCharacteristic(int id)`.
2) I propose `SimulatedPeripheral.getCharacteristic(int id)`.

Former extracts outer loop for reuse in write/monitor, latter extracts inner loop for the same purposes.

Reply:
Extracted.